infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java

```diff
@@ -113,25 +119,42 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
 
       //TODO Rework it to parallel waiting
       for (OpenShiftMachine machine : machines.values()) {
-        machine.waitRunning(machineStartTimeoutMin);
         final String machineName = machine.getName();
-        bootstrapperFactory
-            .create(
-                getContext().getIdentity(),
-                getContext().getEnvironment().getMachines().get(machineName).getInstallers(),
-                machine)
-            .bootstrap();
-
-        ServersReadinessChecker check =
-            new ServersReadinessChecker(machineName, machine.getServers(), serverCheckerFactory);
-        check.startAsync(new ServerReadinessHandler(machineName));
-        check.await();
-        sendRunningEvent(machine.getName());
+        try {
+          machine.waitRunning(machineStartTimeoutMin);
+          bootstrapperFactory
+              .create(
+                  getContext().getIdentity(),
+                  getContext().getEnvironment().getMachines().get(machineName).getInstallers(),
+                  machine)
+              .bootstrap();
+          final ServersReadinessChecker check =
+              new ServersReadinessChecker(machineName, machine.getServers(), serverCheckerFactory);
+          check.startAsync(new ServerReadinessHandler(machineName));
+          check.await();
+          sendRunningEvent(machine.getName());
+        } catch (InfrastructureException rethrow) {
+          sendFailedEvent(machineName, rethrow.getMessage());
+          throw rethrow;
+        }
+      }
+    } catch (InfrastructureException | RuntimeException | InterruptedException e) {
```

Comment:
This method is already complex enough. I think it is a good point to refactor it to reduce complexity. Can you do this in current PR?